File: app/src/test/java/eu/sesma/dagger/NavigatorShould.java
package eu.sesma.dagger;

import eu.sesma.dagger.ui.ActCollaborator;
import eu.sesma.dagger.ui.BaseActivity;
import eu.sesma.dagger.ui.IAppCollaborator;
import eu.sesma.dagger.ui.Navigator;
import eu.sesma.dagger.ui.another.AnotherActivity;
import eu.sesma.dagger.ui.detail.DetailActivity;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import android.content.Context;
import android.content.Intent;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.doNothing;

@RunWith(RobolectricTestRunner.class)
public class NavigatorShould {

    @Mock
    Context context;
    @Mock
    IAppCollaborator appCollaborator;
    @Mock
    ActCollaborator actCollaborator;
    @Mock
    BaseActivity baseActivity;

    @Captor
    ArgumentCaptor<Intent> intentArgumentCaptor;

    private Navigator navigator;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        navigator = new Navigator(context, appCollaborator, actCollaborator, baseActivity);
        doNothing().when(baseActivity).startActivity(intentArgumentCaptor.capture());
    }

    @Test
    public void navigateToDetailOnMethodCall() {

        navigator.navigateToDetail();

        Intent intent = intentArgumentCaptor.getValue();
        assertThat(intentArgumentCaptor.getValue().getComponent().getClassName())
                .isEqualTo(DetailActivity.class.getName());
    }

    @Test
    public void navigateToAnotherOnMethodCall() {

        navigator.navigateToAnother();

        Intent intent = intentArgumentCaptor.getValue();
        assertThat(intentArgumentCaptor.getValue().getComponent().getClassName())
                .isEqualTo(AnotherActivity.class.getName());
    }
}

File: app/src/test/java/eu/sesma/dagger/main/NavigatorShould.kt
package eu.sesma.dagger.main

import android.content.Context
import android.content.Intent
import android.support.v7.app.AppCompatActivity
import com.nhaarman.mockito_kotlin.doNothing
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import eu.sesma.dagger.core.CoreCollaborator
import eu.sesma.dagger.core.CoreSingletonCollaborator
import eu.sesma.dagger.detail.DetailActivity
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class NavigatorShould {

    private val context: Context = mock()
    private val coreCollaborator: CoreCollaborator = mock()
    private val coreSingletonCollaborator: CoreSingletonCollaborator = mock()
    private val mainCollaborator: MainCollaborator = mock()
    private val mainScopedCollaborator: MainScopedCollaborator = mock()
    private val activity: AppCompatActivity = mock()


    private val intentArgumentCaptor = ArgumentCaptor.forClass(Intent::class.java)

    private lateinit var navigator: MainNavigator

    @Before
    fun setup() {
        navigator = MainNavigator(context, coreCollaborator, coreSingletonCollaborator, mainCollaborator, mainScopedCollaborator, activity)
        doNothing().whenever(activity).startActivity(intentArgumentCaptor.capture())
    }

    @Test
    fun navigateToDetailOnMethodCall() {

        navigator.navigateToDetail()

        assertEquals(DetailActivity::class.java.name, intentArgumentCaptor.value.component?.className)
    }
}